src/plonkish/compiler/mod.rs

```diff
@@ -88,7 +90,7 @@ pub fn compile_phase1<
             unit.columns.clone(),
             unit.placement.clone(),
             unit.selector.clone(),
-            TraceGenerator::new(Rc::clone(v), ast.num_steps),
+            TG::new(Rc::clone(v), ast.num_steps),
```

Comment:
Do I get it right that here the type of TG depends on the caller of `compile_phase1`? This is because `compile_phase1` is called from both python frontend and DSL. DSL is calling it via [`sub_circuit`](https://github.com/privacy-scaling-explorations/chiquito/blob/81cdc04991c776b498d0f1ee3704bff2490793ea/src/frontend/dsl/sc.rs#L45), python frontend is calling it via [`sub_circuit_with_ast`](https://github.com/privacy-scaling-explorations/chiquito/blob/81cdc04991c776b498d0f1ee3704bff2490793ea/src/frontend/dsl/sc.rs#L80). In [`chiquito_super_circuit_halo2_mock_prover`](https://github.com/privacy-scaling-explorations/chiquito/blob/81cdc04991c776b498d0f1ee3704bff2490793ea/src/frontend/python/mod.rs#L108) pythoin frontend is deserializing witness and calling [`map_with_witness`](https://github.com/privacy-scaling-explorations/chiquito/blob/81cdc04991c776b498d0f1ee3704bff2490793ea/src/plonkish/ir/sc.rs#L112). As I understand, you want to get rid of that too because the `TraceGenerator` should know where the witness is coming from. However, the witness was obtained [here from the `super_witness`](https://github.com/privacy-scaling-explorations/chiquito/blob/81cdc04991c776b498d0f1ee3704bff2490793ea/src/frontend/python/mod.rs#L132-L136) after both AssignmentGenerator and TraceGenerator were created. Does this mean that I need to refactor their creation and somehow pass the supercircuit witness JSON into them? E.g., by passing `super_witness` into `sub_circuit_with_ast` and `compile_phase1`?

Reply:
@leolara I understand that naturally "new" feels like a constructor and shouldn't be in the trait, but this problem of who is instantiating the TraceGenerator is still unresolved. It is generic here so I could only put the "new" in the trait.

Reply:
@leolara I think I see one possible solution for this. What if we have a TraceGeneratorFactory "abstract factory" trait? Then we can pass the concrete implementations from the outside of the `sub_circuit -> compile_phase1` chain. These concrete implementations will know how to instantiate the corresponding `TraceGenerator` with or without arguments.

Reply:
Let me send a commit, What we just need is to have in SBPIR the TraceGenerator as the field, instead of the closure